nncf/experimental/torch/fracbits/quantizer.py

```diff
@@ -0,0 +1,222 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+# Reference: Yang, Linjie, and Qing Jin. "Fracbits: Mixed precision quantization via fractional bit-widths."
+# Proceedings of the AAAI Conference on Artificial Intelligence. Vol. 35. No. 12. 2021.
+
+from typing import Dict
+import torch
+
+from nncf.experimental.torch.fracbits.structs import FracBitsQuantizationMode
+
+from nncf.torch.layer_utils import COMPRESSION_MODULES, CompressionParameter
+from nncf.torch.quantization.layers import (
+    QUANTIZATION_MODULES, AsymmetricQuantizer, PTQuantizerSpec, SymmetricQuantizer)
+from nncf.torch.quantization.quantize_functions import asymmetric_quantize, symmetric_quantize
+from nncf.torch.utils import no_jit_trace
+
+
+@COMPRESSION_MODULES.register()
+@QUANTIZATION_MODULES.register(FracBitsQuantizationMode.SYMMETRIC)
+class FracBitsSymmetricQuantizer(SymmetricQuantizer):
+    def __init__(self, qspec: PTQuantizerSpec):
+        self._constructed = False
+        super().__init__(qspec)
+        self._min_num_bits = int(0.5 * qspec.num_bits)
+        self._max_num_bits = int(1.5 * qspec.num_bits)
+        self._num_bits = CompressionParameter(torch.FloatTensor([qspec.num_bits]), requires_grad=True,
+                                              compression_lr_multiplier=qspec.compression_lr_multiplier)
+        self._constructed = True
+
+    @property
+    def frac_num_bits(self):
+        return torch.clamp(self._num_bits, self._min_num_bits, self._max_num_bits)
+
+    @property
+    def num_bits(self):
+        if not self._constructed:
```

Comment:
This is because at construction time `super().__init__()` is called and at that time `_num_bits` is an integer tensor. 
https://github.com/openvinotoolkit/nncf/blob/13cc15e6560985606abfd55086242d9f7687d6dd/nncf/torch/quantization/layers.py#L272-L273
During `super().__init__()`, `self.num_bits` should be called but `_num_bits.floor()` is not allowed because it is an integer tensor. To handle this case I made a flag `self._constructed`. But I think we can remove this flag and check it like `self._num_bits.dtype == torch.int`.

Reply:
Fixed.

Reply:
I asked that there is any case of `self._constructed = False`. It was always set to `True` when the class is declared.

Reply:
It was
```
1. self._constructed = False
2. super().__init__()
    (During 2.) Call self.num_bits with self._constructed = False
3. self._constructed = True
4. Can call self.num_bits with self._constructed = True
```

Reply:
`self._constructed` could be `False` during `__init__` of the base class (`BaseQuantizer`).